### PR TITLE
Add @trace decorator to log traces for sub functions in tools

### PR DESCRIPTION
**Pull request by zzn2**
# Description
Currently, the traces of tools and LLM function calls are automatically generated.

This PR is to make it possible to add traces for any Python functions that are called by tools.
To generate a trace for a function, simply add `@trace` decorator to it.

Here is a sample:
```diff
+ @trace
  def is_valid_name(name):
      sleep(0.5)
      return len(name) > 0


+ @trace
  def get_user_name(user_id):
      sleep(0.5)
      user_name = f"User {user_id}"
      if not is_valid_name(user_name):
          raise ValueError(f"Invalid user name: {user_name}")

      return user_name


+ @trace
  def format_greeting(user_name):
      sleep(0.5)
      return f"Hello, {user_name}!"


  @tool
  def greetings(user_id):
      user_name = get_user_name(user_id)
      greeting = format_greeting(user_name) 
      print(greeting)
      return {"greeting": greeting}
```
Here is what the traces look like:

Before adding `@trace` decorators:
image

After adding `@trace` decorators:
image

# Copilot-generated Description
This pull request includes a variety of changes to improve the tracing functionality and testing of the codebase. The most important changes include adding a new function to enable tracing for both synchronous and asynchronous functions, adding a new test case to verify the behavior of flows with trace marks, and modifying the `tool.py` file to improve the implementation of the `tool_decorator` function.

Tracing improvements:

* <a href="diffhunk://#diff-8f8c2ae53e5ffd37a14e8a899119fbb2742486db8faab6df3fcf506e1b720ad8R153-R258">`src/promptflow/promptflow/_core/tracer.py`</a>: Added a new function `_traced` to enable tracing for both synchronous and asynchronous functions.
* <a href="diffhunk://#diff-a36689a949893c227689d9dadb4c1e7008a0551206871e371658f4cc70f16f07R151-R219">`src/promptflow/tests/executor/e2etests/test_traces.py`</a>: Added a new test case to verify the behavior of flows with trace marks.
* <a href="diffhunk://#diff-4d476f676cf919f26dea6de37d8ec9136420d68cd8f4b0d376471496251a3718L5-R14">`src/promptflow/promptflow/_core/tool.py`</a>: Modified the `tool.py` file to improve the implementation of the `tool_decorator` function.

Testing improvements:

* <a href="diffhunk://#diff-2c3a4087454f99dd0ef273a55b0847c70887a4be33ec66e2329c8908ecb44bd5L122-R126">`src/promptflow/tests/sdk_cli_test/recording_utilities/mock_tool.py`</a>: Modified the `mock_tool.py` file to automatically trace tool functions and remove unnecessary imports.
* <a href="diffhunk://#diff-a59cea65d9eaf47ebc764141c09ecd84c35a38a332218cdc7bd14e2dd2f66196R1-R32">`src/promptflow/tests/test_configs/flows/flow_with_trace/greetings.py`</a>: Introduced tracing to the greetings flow by adding a new Python file with decorated functions and a tool function.
* <a href="diffhunk://#diff-d572bf03138454bb6cf9229e9ea52580360c274c94e329d60e7fd485e6c41578R127-R130">`src/promptflow/tests/executor/unittests/_core/test_tracer.py`</a>: Added additional tests to verify trace creation, attribute setting, and error handling for invalid tool types.

Other changes:

* <a href="diffhunk://#diff-d12fdd7b90cc1748f1d3e1237b4f357ba7f66740445d117beeb68ed174d1e86eL170-R179">`src/promptflow/promptflow/_utils/utils.py`</a>: Updated the import statement in the `extract_user_frame_summaries` function to fix an import error.
* <a href="diffhunk://#diff-893908f9e17225cd374df3db5f43b1131e3e51c55f25bc6a48d95d095e5b8550R14">`src/promptflow/dev_requirements.txt`</a>: Added the `pytest-asyncio` package to the development requirements.
* <a href="diffhunk://#diff-6f6440cf607ad46fd929835113505676abd8faf84dbfe285f6d210dcc69bdfb2R10">`src/promptflow/promptflow/__init__.py`</a>: Imported the `trace` function and added it to the `__all__` list.
* <a href="diffhunk://#diff-025f6ace207470504d15aeceb310625c6178fca578f856e0f6e36dd1a7f7b48eR13">`src/promptflow/promptflow/contracts/trace.py`</a>: Added a new trace type to the `TraceType` enumeration class.

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_core/tool.py [hunk 1]
```diff
@@ -133,6 +133,95 @@ def decorated_tool(*args, **kwargs):
     return tool_decorator
 
 
+def trace(
```
Comment (D-W-) [inline src/promptflow/promptflow/_core/tool.py:136]: COOL! Will @tool call this?

File: src/promptflow/promptflow/_core/tool.py [hunk 2]
```diff
@@ -133,6 +133,95 @@ def decorated_tool(*args, **kwargs):
...
+                except Exception as e:
+                    Tracer.pop(None, e)
+                    raise
+
+            new_f = decorated_tool
+
+        if type is not None and type not in [k.value for k in ToolType]:
+            raise UserErrorException(f"Tool type {type} is not supported yet.")
+
+        new_f.__original_function = func
```
Comment (thy09) [inline src/promptflow/promptflow/_core/tool.py:207]: All following logic are tool specific, I think we just need L160 to L200

Comment (zzn2) [reply src/promptflow/promptflow/_core/tool.py:207]: Sure. Will remove them.

File: src/promptflow/promptflow/_utils/utils.py [hunk 3]
```diff
@@ -167,17 +167,16 @@ def log_progress(
...
+    core_folder = os.path.dirname(_core.__file__)
 
     for i in range(len(frame_summaries) - 1):
         cur_file = frame_summaries[i].filename
         next_file = frame_summaries[i + 1].filename
-        # If the current frame is in tool.py and the next frame is not in _core folder
+        # If the current frame is in _core folder and the next frame is not in _core folder
         # then we can say that the next frame is in user code.
-        if cur_file == tool_file and not next_file.startswith(core_folder):
+        if cur_file.startswith(core_folder) and not next_file.startswith(core_folder):
```
Comment (zzn2) [inline src/promptflow/promptflow/_utils/utils.py:179]: The entry point to call tool has changed from tool.py to tracer.py.
They are both in `_core` folder, so I updated the checks here accordingly.
(Confirmed that the count of frames has not changed).

File: src/promptflow/promptflow/_core/tracer.py [hunk 4]
```diff
@@ -156,3 +150,109 @@ def _create_trace_from_function_call(f, *, args=[], kwargs={}, trace_type=TraceT
...
+        async def greetings_async(user_id):
+            name = await get_name_async(user_id)
+            return f"Hello, {name}"
+    """
+
+    def wrapper(func):
+        return _traced(func, trace_type=TraceType.FUNCTION)
+
+    # enable use decorator without "()" if all arguments are default values
+    if func is not None:
```
Comment (thy09) [inline src/promptflow/promptflow/_core/tracer.py:256]: We may not support `trace()` for now.

**Review by thy09 — APPROVED** [review]

File: src/promptflow/promptflow/_core/tracer.py (hunk 4 above)
Comment (zzn2) [reply src/promptflow/promptflow/_core/tracer.py]: Sure. Removed.

**Review by D-W- — APPROVED** [review]

**Pull request merged**